File: PVSupport/PVSupport/NSObject+PVAbstractAdditions.m
//
//  NSObject+PVAbstractAdditions.m
//  Provenance
//
//  Created by James Addyman on 31/08/2013.
//  Copyright (c) 2013 James Addyman. All rights reserved.
//

#import "NSObject+PVAbstractAdditions.h"
#import "PVLogging.h"

@implementation NSObject (PVAbstractAdditions)

+ (void)doesNotImplementSelector:(SEL)aSel
{
    @throw [NSException exceptionWithName:NSInvalidArgumentException
                                   reason:[NSString stringWithFormat:@"*** +%s cannot be sent to the abstract class %@: Create a concrete subclass!", sel_getName(aSel), [self class]]
                                 userInfo:nil];
}

- (void)doesNotImplementSelector:(SEL)aSel
{
    @throw [NSException exceptionWithName:NSInvalidArgumentException
                                   reason:[NSString stringWithFormat:@"*** -%s cannot be sent to an abstract object of class %@: Create a concrete instance!", sel_getName(aSel), [self class]]
                                 userInfo:nil];
}

+ (void)doesNotImplementOptionalSelector:(SEL)aSel
{
    ELOG(@"*** +%s is an optional method and it is not implemented in %@!", sel_getName(aSel), NSStringFromClass([self class]));
}

- (void)doesNotImplementOptionalSelector:(SEL)aSel
{
    ELOG(@"*** -%s is an optional method and it is not implemented in %@!", sel_getName(aSel), NSStringFromClass([self class]));
}


@end

File: PVSupport/Sources/PVSupport/NSExtensions/NSObject+PVAbstractAdditions.swift
//
//  NSObject+PVAbstractAdditions.swift
//  PVSupport
//
//  Created by Joseph Mattiello on 8/10/22.
//  Copyright © 2022 Provenance Emu. All rights reserved.
//

import Foundation

@objc
public extension NSObject {
    class func doesNotImplementSelector(_ sel: Selector) {
        let exception = NSException(name: .invalidArgumentException, reason: "*** \(sel_getName(sel)) cannot be sent to the abstract class \(type(of: self)): Create a concrete subclass!")
        objc_exception_throw(exception)
    }
    
    func doesNotImplementSelector(_ sel: Selector) {
        let exception = NSException(name: .invalidArgumentException, reason: "*** \(sel_getName(sel)) cannot be sent to the abstract class \(type(of: self)): Create a concrete subclass!")
        objc_exception_throw(exception)
    }
    
    
    class func doesNotImplementOptionalSelector(_ sel: Selector) {
        ELOG("*** + \(sel_getName(sel)) is an optional method and it is not implemented in \(type(of: self))!")
    }
    
    func doesNotImplementOptionalSelector(_ sel: Selector) {
        ELOG("*** + \(sel_getName(sel)) is an optional method and it is not implemented in \(type(of: self))!")
    }
}
